cpp/src/arrow/dataset/filter.cc

```diff
@@ -758,43 +797,107 @@ Result<std::shared_ptr<DataType>> FieldExpression::Validate(const Schema& schema
   return null();
 }
 
-struct AccumulateFieldsVisitor {
-  void operator()(const FieldExpression& expr) { fields.push_back(expr.name()); }
+Result<std::shared_ptr<Expression>> InsertImplicitCasts(const Expression& expr,
+                                                        const Schema& schema) {
+  struct {
+    struct Validated {
+      std::shared_ptr<Expression> expr;
+      std::shared_ptr<DataType> type;
+    };
+
+    Result<Validated> Validate(const Expression& expr) {
+      Validated out;
+      ARROW_ASSIGN_OR_RAISE(out.expr, InsertImplicitCasts(expr, schema_));
+      ARROW_ASSIGN_OR_RAISE(out.type, out.expr->Validate(schema_));
+      return std::move(out);
```

Comment:
Is `std::move` actually necessary? Is C++ able to devise that a return variable is a rvalue?

Reply:
You're right and this *should* be unnecessary.
I placed it here to silence warnings from clang that older compilers may not infer implicit move because the return type is not identical to the variable type